### ISSUE TITLE
`no-throw-literal` does not catch concatenated exception messages

### ISSUE THREAD

**Issue by feross** [eslint/eslint#3099]
Using eslint `master`, and the following code:

```
throw 'a' + 'b'
```

and the following rule:

```
"no-throw-literal": 2
```

I expect an error, but get none.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by ilyavolodin** [comment]
This is reproducible in online demo.

**Comment by IanVS** [comment]
Working on it.

**Comment by IanVS** [comment]
[This commit](https://github.com/taromero/latte/commit/924200626589743b3056f9c1b5b987c4f5d6bf04) appears to be what precipitated this issue report.

The problem with trying to catch something like this would be testing for the following:

```js
var err = new Error("error");
throw 'Now I am throwing an ' + err;
```

The above throws an actual Error, and should be considered valid.  Whereas:

```js
var err = 'literal';
throw 'Now I am throwing an ' + err;
```

ends up throwing a literal and ideally would be flagged.  However, I don't see a good way to handle this via static analysis.  Anyone else is welcome to try, but I have to back out I'm afraid.

**Comment by feross** [comment]
Does your first example actually throw an Error? Looks like it throws a string to me.

**Comment by IanVS** [comment]
Hm, looks like you're right.  I get:

```js
throw 'Now I am throwing an ' + err;
                              ^
Now I am throwing an Error: error
```

Which at first glance looked like an error but isn't, the error was recast to a string.

So, maybe one solution would be to catch any `throw` with a `BinaryExpression` argument, since it will not result in a real error?

**Comment by mysticatea** [comment]
Yes, I guess this rule should catch below:

- `ArrayExpression` (an array literal)
- `FunctionExpression` (a function literal)
- `ArrowFunctionExpression`
- `ClassExpression` (a class literal)
- `TemplateElement` (a string literal)
- `UnaryExpression` (converted implicitly to boolean, number, string, or undefined)
- `BinaryExpression` (converted implicitly to boolean, number, or string)
- `UpdateExpression` (converted implicitly to number)

And should catch partially below:

- `SequenceExpression` - checks its last element of `expressions` recursively.
- `LogicalExpression` - checks both `left` and `right` recursively.
- `ConditionalExpression` - checks both `alternate` and `consequent` recursively.

However, this probably be the overspec :)
